### PR TITLE
RIRPA AGW updates

### PR DESCRIPTION
**Pull request by cjcscott**
This updates the RIRPA functionality within Vayesta to be consistent with the discussions of error estimation within the [AGW preprint](https://arxiv.org/abs/2301.09107) along with some optimisations for practical usage.
Most changes are to `ssRIRPA.kernel_moms`, specifically:
- Analytic bounds on the error are only calculated if the option `analytic_lower_bound=True`; this calculation costs O[N^4] and since this is not a tight lower or upper bound it's not very useful compared to the (cheaper) approximate upper bound which we always calculate from Curtis-Clenshaw quadrature.
- By default the auxiliary indices for low-rank representations are no longer compressed via SVD; this is controlled via the object attribute `ssRIRPA.compress` (larger values result in ensuring the most compact representations of more matrices, defaults to 0). Since compression requires O[N_{aux}^2 N^2] computation it should only be used when it's likely to make a considerable reduction our dominant computational cost- the construction of Q at each NI point, at O[n_{points}N_{aux}^2 N^2]. For practical numbers of points compression isn't currently worth it.
- Added full docstring of all options for `.kernel_moms`.
- Added regression testing for ssRIRPA error estimates.
- Added option to return low-level `NumericalIntegratorBase` objects from calculation for analysis of integrand values.

All changes are non-breaking and change no exposed interfaces, so should just work with any existing scripts.

### CODE REVIEW

**Comment by cjcscott** [comment]
I should also note that altogether for a quick calculation on STO-3G icosane (C20H42) with 20 integration points these changes reduce the cost of computing the full zeroth moment by a bit more than a factor of 2. Profiling suggests ~80% of the runtime for the moment calculation is now computing and inverting Q, as expected (the remaining 20% is the O[N^4] computation and application of (A+B)^{-1}, which will be replaced by an O[N^3] iterative approach in an update following shortly).

**Review by maxnus — COMMENTED** [review]
Some code style comments

File: vayesta/rpa/rirpa/RIRPA.py [hunk 1]
```diff
@@ -130,11 +183,35 @@ def kernel_moms(self, max_moment, target_rot=None, npoints=48, ainit=10, integra
             Dsq = D ** 2
             for i in range(2, max_moment + 1):
                 moments[i] = einsum("pq,q->pq", moments[i - 2], Dsq) + dot(moments[i - 2], ri_mp[1].T, ri_mp[0])
```
[the remark this answers is not in the document]
Comment (cjcscott) [reply vayesta/rpa/rirpa/RIRPA.py]: `self.D` constructs D, which has some overhead, hence keeping it around locally if it's being reused immediately

File: vayesta/rpa/rirpa/NI_eval.py [hunk 2]
```diff
@@ -273,9 +273,14 @@ def _NI_eval_w_error(self, a, res_shape, evaluator):
...
-        # error = b ** 3 / a ** 2
-        # error_error = b ** 2 / (a ** 2 + b ** 2)
+        #error = b ** 3 / a ** 2
+        #error_error = b ** 2 / a ** 2
+
+        #if error_error > self.linear_error_tol:
+        #    self.log.info("Error in linear approx for eta0 error exceeded %4.2f (%4.2f); falling back to cubic model.",
+        #                  self.linear_error_tol, error_error)
         error = self.calculate_error(a, b)
+        #error = self.calculate_error(a, b)
```
[the remark this answers is not in the document]
Comment (cjcscott) [reply vayesta/rpa/rirpa/NI_eval.py]: Reworked into a more descriptive comment- this is just a simpler approximation which makes understanding the more complicated approach easier (for me at least).

File: vayesta/rpa/rirpa/RIRPA.py [hunk 3]
```diff
@@ -108,8 +117,52 @@ def D(self):
...
+            max_moment: integer
+                Maximum moment of the dd response to return.
+            target_rot: np.array, of size (n_{target}, o_a v_a + o_b v_b)
+                Projector for one index of the moment.
+            npoints: int, optional.
+                Integer number of points to use in numerical integrations; will be increased to next highest multiple of
+                4 for error estimation purposes. Default: 48 (excessive).
+            integral_deduct: str, optional.
+                What terms to deduct from numerical integration.
+                Options are "HO" (default), "D", and None, corresponding to deducting both the mean-field contribution
```
[the remark this answers is not in the document]
Comment (cjcscott) [reply vayesta/rpa/rirpa/RIRPA.py]: I've added a pointer to the AGW preprint (https://arxiv.org/abs/2301.09107) for details of what's being done here; the only abbreviation is **H**igher-**O**rder, D just corresponds to just deducting terms related to the matrix **D**

**Pull request merged**